walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -20,27 +20,52 @@
 import java.util.Collections;
 import java.util.LinkedList;
 import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
 
 public final class FindMeetingQuery {
-  private final Collection<TimeRange> mAvailableTimes = new ArrayList<>();
-  private long mMeetingDuration;
-
-  /* Returns all available slots for a meeting to happen
+  /*
+   * 1. If one or more time slots exists so that both mandatory and optional attendees can attend,
+   *                                              returns those time slots.
+   *
+   * 2. Otherwise, returns the time slots that fit just the mandatory attendees.
+   *
    * Parameters: events (all events that are in the calendar, some of them could not relate
    *                      to the attendees in the meeting request) */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    if (request.getAttendees().isEmpty()) {
-      return Arrays.asList(TimeRange.WHOLE_DAY);
-    }
-
     if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
       return Arrays.asList();
     }
 
+    Collection<String> mandatoryAttendees = request.getAttendees();
+    Collection<String> optionalAttendees = request.getOptionalAttendees();
+    long meetingDuration = request.getDuration();
+
+    if (optionalAttendees.isEmpty()) {
+      return getAvailableTimes(mandatoryAttendees, events, meetingDuration);
+    }
+
+    Collection<String> allAttendees = Stream.
+            concat(mandatoryAttendees.stream(), optionalAttendees.stream()).
+            collect(Collectors.toList());
+
+    Collection<TimeRange> availableTimesForAllAttendees = getAvailableTimes(allAttendees, events, meetingDuration);
```

Comment:
```suggestion
    Collection<TimeRange> availableTimesForAllAttendees = getAvailableTimes(
        Stream.concat(mandatoryAttendees.stream(), optionalAttendees.stream())
            .collect(Collectors.toList()),
        events,
        meetingDuration);
```


Suggestion to inline `allAttendees` since it's only used once. This isn't necessarily better so feel free to leave as is.